internal/cri/server/images/image_pull.go

```diff
@@ -266,6 +240,105 @@ func (c *CRIImageService) PullImage(ctx context.Context, name string, credential
 	return imageID, nil
 }
 
+// pullImageWithLocalPull handles image pulling using the local client.
+func (c *CRIImageService) pullImageWithLocalPull(
+	ctx context.Context,
+	ref string,
+	credentials func(string) (string, string, error),
+	snapshotter string,
+	labels map[string]string,
+	imagePullProgressTimeout time.Duration,
+	imageHandler containerdimages.HandlerFunc,
+) (containerd.Image, error) {
+	pctx, pcancel := context.WithCancel(ctx)
+	defer pcancel()
+	pullReporter := newPullProgressReporter(ref, pcancel, imagePullProgressTimeout)
+	resolver := docker.NewResolver(docker.ResolverOptions{
+		Headers: c.config.Registry.Headers,
+		Hosts:   c.registryHosts(ctx, credentials, pullReporter.optionUpdateClient),
+	})
+
+	log.G(ctx).Debugf("PullImage %q with snapshotter %s using client.Pull()", ref, snapshotter)
+	pullOpts := []containerd.RemoteOpt{
+		containerd.WithSchema1Conversion, //nolint:staticcheck // Ignore SA1019. Need to keep deprecated package for compatibility.
+		containerd.WithResolver(resolver),
+		containerd.WithPullSnapshotter(snapshotter),
+		containerd.WithPullUnpack,
+		containerd.WithPullLabels(labels),
+		containerd.WithMaxConcurrentDownloads(c.config.MaxConcurrentDownloads),
+		containerd.WithImageHandler(imageHandler),
+		containerd.WithUnpackOpts([]containerd.UnpackOpt{
+			containerd.WithUnpackDuplicationSuppressor(c.unpackDuplicationSuppressor),
+			containerd.WithUnpackApplyOpts(diff.WithSyncFs(c.config.ImagePullWithSyncFs)),
+		}),
+	}
+
+	// Temporarily removed for v2 upgrade
+	//pullOpts = append(pullOpts, c.encryptedImagesPullOpts()...)
+	if !c.config.DisableSnapshotAnnotations {
+		pullOpts = append(pullOpts,
+			containerd.WithImageHandlerWrapper(snpkg.AppendInfoHandlerWrapper(ref)))
+	}
+
+	if c.config.DiscardUnpackedLayers {
+		// Allows GC to clean layers up from the content store after unpacking
+		pullOpts = append(pullOpts,
+			containerd.WithChildLabelMap(containerdimages.ChildGCLabelsFilterLayers))
+	}
+
+	pullReporter.start(pctx)
+	image, err := c.client.Pull(pctx, ref, pullOpts...)
+	pcancel()
+	if err != nil {
+		return nil, fmt.Errorf("failed to pull and unpack image %q: %w", ref, err)
+	}
+	return image, nil
+}
+
+// pullImageWithTransferService handles image pulling using the transfer service.
+func (c *CRIImageService) pullImageWithTransferService(
+	ctx context.Context,
+	ref string,
+	credentials func(string) (string, string, error),
+	snapshotter string,
```

Comment:
Does this support remote snapshotter? Seems transfer service doesn't support some remote snapshotter based on https://github.com/containerd/containerd/issues/9931. Can we add a comment/TODO in the comment?

Reply:
I haven't tried running remote snapshotters with the transfer service. But looks like the issue is still open. I will try running it with a remote snapshotter and if there are more issues with CRI. Also agree that we can add a TODO comment here and follow up on the open issue. Thanks for pointing this out.

Reply:
FYI  #11195 and #11196 try to solve this